src/runtime_src/xdp/profile/plugin/aie_debug/aie_debug_plugin.cpp

```diff
@@ -0,0 +1,397 @@
+          if (coreUnstalled && graphStallCounter) {
+            graphStallCounter--;
+          } else if (coreStallCounter == GRAPH_HANG_COUNT_THRESHOLD) {
+            graphStallCounter++;
+          }
+
+          // Is this core stuck for long time?
+          if (coreStallCounter == CORE_HANG_COUNT_THRESHOLD) {
+            foundStuckCores = true;
+            stuckTile = tile;
+            stuckCoreStatus = coreStatus;
+          }
+
+          coreStatusMap[tile] = coreStatus;
+        } // For tiles in graph
+
+        std::stringstream warningMessage;
+        if (graphStallCounter == graphTilesVec.size()) {
+          // We have a stuck graph
+          warningMessage
+          << "Potential deadlock/hang found in AI Engines. Graph : " << graphName;
+
+          xrt_core::message::send(severity_level::warning, "XRT", warningMessage.str());
+          // Send next warning if all tiles come out of hang & reach threshold again
+          graphStallCounter = 0;
+        } else if (foundStuckCores) {
+          // We have a stuck core within this graph
+          warningMessage
+          << "Potential deadlock/hang found in AI Engines. Graph : " << graphName << " "
+          << "Tile : " << "(" << stuckTile.col << "," << stuckTile.row + 1 << ") "
+          << "Status 0x" << std::hex << stuckCoreStatus << std::dec
+          << " : " << getCoreStatusString(stuckCoreStatus);
+
+          xrt_core::message::send(severity_level::warning, "XRT", warningMessage.str());
+          foundStuckCores = false;
+        }
+
+        // Print status for debug
+        if (xrt_core::config::get_verbosity() >= static_cast<unsigned int>(severity_level::debug)) {
+          std::stringstream msg;
+          for (const auto& tile : graphTilesVec) {
+            if (coreStuckCountMap[tile]) {
+              msg
+                << "T(" << tile.col <<"," << tile.row + 1 << "):" << "<" << coreStuckCountMap[tile]
+                << ":0x" << std::hex << coreStatusMap[tile] << std::dec << "> ";
+            }
+          }
+          if (!msg.str().empty()) {
+            msg << std::endl << "Graph " << graphName << " #Cur : " << graphStallCounter << " #Thr : " << graphTilesVec.size();
+            xrt_core::message::send(severity_level::debug, "XRT", msg.str());
+          }
+        }
+      } // For graphs
+
+      // Always write out latest debug/status file
+      for (auto w : writers) {
+        w->write(false);
+      }
+
+      std::this_thread::sleep_for(std::chrono::microseconds(mPollingInterval));     
+    }
+  }
+
+  void AIEDebugPlugin::updateAIEDevice(void* handle)
+  {
+    // Don't update if no debug/status is requested
+    if (!xrt_core::config::get_aie_status())
+      return;
+
+    const unsigned int PATH_LENGTH = 512;
+    char pathBuf[PATH_LENGTH];
+    memset(pathBuf, 0, PATH_LENGTH);
+    xclGetDebugIPlayoutPath(handle, pathBuf, PATH_LENGTH);
+
+    std::string sysfspath(pathBuf);
```

Comment:
This looks like some very brittle code.
At least this should be a library function.
Why 512?
What happens if more than 512 characters is needed?

Reply:
Good catch. sysfs names are't too long as of now. This should be fine but needs to be reviewd in future for all the plugins.

Reply:
What if someone is fuzzing your program to detect some stack overflow?
Do you trust `sysfs` inside the Matrix? :-)
I do not know the detail of the `xclGetDebugIPlayoutPath` API.
Perhaps at least `xclGetDebugIPlayoutPath(handle, pathBuf, PATH_LENGTH - 1);` is a cheap way to remove a security bug? @stsoe